index.js

```diff
@@ -1,96 +1,98 @@
+  return AtomicStore.prototype.createWriteStream.call(this, opts, function (err, metadata) {
+    metadata.key = originalKey
+    cb(err, metadata)
   })
 }
 
-BlobStore.prototype.createReadStream = function (opts) {
-  var name = typeof opts === 'string' ? opts : opts.key
-  var subdir = filenamePrefix(name, 7)
-  var store = this._getStore(subdir)
-  return store.createReadStream(opts)
+BlobStore.prototype.createReadStream = function (key) {
+  if (key && typeof key === 'object') return this.createReadStream(key.key)
+  key = this._insertSubDirPrefix(key)
+  return AtomicStore.prototype.createReadStream.call(this, key)
 }
 
-BlobStore.prototype.exists = function (opts, done) {
-  var name = typeof opts === 'string' ? opts : opts.key
-  var subdir = filenamePrefix(name, 7)
-  var store = this._getStore(subdir)
-  return store.exists(opts, done)
+BlobStore.prototype.exists = function (opts, cb) {
+  if (typeof opts === 'string') opts = {key: opts}
+  opts.key = this._insertSubDirPrefix(opts.key)
+  return AtomicStore.prototype.exists.call(this, opts, cb)
 }
 
-BlobStore.prototype.remove = function (opts, done) {
-  var name = typeof opts === 'string' ? opts : opts.key
-  var subdir = filenamePrefix(name, 7)
-  var store = this._getStore(subdir)
-  return store.remove(opts, done)
+BlobStore.prototype.remove = function (opts, cb) {
+  if (typeof opts === 'string') opts = {key: opts}
+  opts.key = this._insertSubDirPrefix(opts.key)
+  return AtomicStore.prototype.remove.call(this, opts, cb)
 }
 
-// TODO: opts to choose whether to use staging area
-BlobStore.prototype.createWriteStream = function (opts, done) {
+BlobStore.prototype.list = function (cb) {
+  var names = []
   var self = this
-  done = done || noop
-
-  var name = typeof opts === 'string' ? opts : (opts.name ? opts.name : opts.key)
-
-  var stagingStore = this._getStore('staging')
-  var ws = stagingStore.createWriteStream(opts)
-  ws.on('finish', onFinish)
-  ws.on('error', onFinish)
-  return ws
-
-  function onFinish (err) {
-    if (err) return done(err)
-
-    var subdir = filenamePrefix(name, 7)
-
-    // write result to destination
-    var from = path.join(self._dir, 'staging', name)
-    var to = path.join(self._dir, subdir, name)
+  walk.files(this.path, function (basedir, filename, stat, next) {
+    var key = path.relative(self.path, path.join(basedir, filename))
+    // Skip tmp files
+    if (key.endsWith(TMP_POSTFIX)) return next()
+    names.push(self._removeSubDirPrefix(key))
+    next()
+  }, function (err) {
+    if (err && err.code === 'ENOENT') cb(null, [])
+    else cb(err, names)
+  })
+}
 
-    mkdirp(path.join(self._dir, subdir), function (err) {
-      if (err) return done(err)
-      fs.rename(from, to, function (err) {
-        done(err, { key: name })
-      })
-    })
-  }
+BlobStore.prototype._insertSubDirPrefix = function (key) {
+  var prefixLen = this.subDirPrefixLen
+  var parsed = path.parse(key)
+  if (parsed.name.length <= prefixLen) return key
```

Comment:
The old code used as much of the string as possible as a prefix. I think that way will give better results than this: not doing any prefixing at all on short keys.

Reply:
I think you can just remove this one line and the `slice` below will work fine on short keys.

Reply:
This will also simplify the removal code below, since you know you can always just pop the top dir off.

Reply:
I think there are edge-cases that could creep in. I changed the way this is done now. There are maybe other ways of doing this but the current way I think I've thought through a few different edge cases.

Reply:
I wrote a breaking test before this commit. The issue is a naming collision between two file named `fo` and `foobar`. Internally these collide (filename with same name as folder). 
The other issue if you are just using the prefix is two files with keys `foobar.txt` and `fo/obar.txt` which would overwrite each other. (This is with a súbdito prefix length of 2)

Reply:
I wrote a breaking test before this commit. The issue is a naming collision between two file named `fo` and `foobar`. Internally these collide (filename with same name as folder). 
The other issue if you are just using the prefix is two files with keys `foobar.txt` and `fo/obar.txt` which would overwrite each other. (This is with a súbdito prefix length of 2)